### ISSUE TITLE
Breaking change - WC_Shipping_Method::add_rate() changed signature

### ISSUE THREAD

**Issue by daigo75** [woothemes/woocommerce#10740]
It seems that method `WC_Shipping_Method::add_rate()` changed signature. This is going to be a breaking change, and shipping methods that use the old signature will generate a warning.

**Method signature in WC 2.5 and earlier**
`WC_Shipping_Method::add_rate( $args = array() )`

**Method signature in WC 2.6**
`WC_Shipping_Method::add_rate( $args = array(), $package = false )`

The new signature doesn't seem necessary. The original method already took a generic array as the only argument, and it would be possible to pass the new `$package` value through it.

Proposed fix
--
Restore original signature and pass the package through the `$args` argument.

**Comment by claudiosmweb** [comment]
"Breaking change" where?

You can keep using `WC_Shipping_Method::add_rate( $args = array() )` as well, since by default `$package` is `false`.

A real breaking change should be something like ` WC_Shipping_Method::add_rate( $args, $package )` that requires pass a `$package`.

**Comment by daigo75** [comment]
I'm fairly surprised by the reply. Changing a signature **is** a potentially breaking change, which throws a "_violation of strict standards_" message. That condition, [from PHP 7 onwards, has been raised in severity to a warning](http://php.net/manual/en/migration70.incompatible.php) (i.e. something that could potentially cause issues and must be addressed).

**Sample error message**
*Strict standards: Declaration of `My_Custom_Shipping::add_rate()` should be compatible with `WC_Shipping_Method::add_rate($args = Array, $package = false)`*

I also noticed that the same issue affects the `WC_Settings_API::process_admin_options()` method.

*Strict standards: Declaration of `My_Custom_Shipping::process_admin_options()` should be compatible with `WC_Settings_API::process_admin_options($post_data = Array)`*

Using default arguments doesn't fix the issue, and it's not a sufficient solution. As mentioned, it should be easy enough to fix the `WC_Shipping_Method::add_rate($args = Array, $package = false)` method by incorporating the package in the `$args` parameter. Unfortunately, this cannot be done with `WC_Settings_API::process_admin_options()`, since it didn't accept any arguments to start with.

If you want to keep the new signatures, I **strongly** recommend to write a prominent note in your development blog, to inform other developers of this change. This will allow them to update their shipping plugins before WC 2.6 is released.

**Comment by mikejolley** [comment]
Notices are warnings are not breaking changes.. don't forget deprecation notices are just that too - notices. If $package were required it would be breaking, but that is not the case.

**Comment by daigo75** [comment]
Perhaps "breaking change" conveys the wrong message, but let's not be too quick to dismiss this issue. A warning is a fairly serious message, it should not be ignored. Just because the warning is not followed by a crash, it doesn't mean that it's unimportant, as it may still result in unexpected behaviour. 

In any case, this kind of changes should be written somewhere, so that other developers won't have to "stumble upon them".

There are dozens of custom shipping plugins on the market, and they will all start throwing warnings in WC 2.6. If they get updated to support WC 2.6, they will start throwing warnings in WC 2.5 and earlier. Supporting WC 2.6 and earlier versions will require a small workaround, but it's still important that developers are aware of it.

**Comment by mikejolley** [comment]
We also have shipping extensions, and yes we'll be updating where required. That doesn't mean that the change is not required.

Also, even though there are likely cases where people don't do it, notices and warnings should be disabled in production.

Fact is, this change is required to support custom shipping meta data. New features require changes..

And yes we have shipping docs (https://github.com/woothemes/woocommerce/wiki/Shipping-Method-API-(2.6)) and we're working on plugin boilerplates for reference.

I'm sure if this were a feature **your** plugins benefitted from you would not mind so much ;)

**Comment by daigo75** [comment]
@mikejolley actually, you are mistaken. My plugins are not affected by those changes. I already designed them to be able to accommodate a fairly good deal of unexpected changes (learned the lesson with WC 2.3 and 2.4), and this won't be an issue in my case. I noticed the message while using 3rd party plugins on a test site, and I simply "_put myself in their developers' shoes_".

Even if the changes were of any benefit to me, I can assure you that we would be here discussing this issue anyway. It should be clear by now that I give backward compatibility a very high priority (almost above everything else), and this kind of changes are not fully backward compatible.

As for "_the changes are required_", I'm sure that there are good reasons for them (I never doubted it). I was simply pointing out that they could be implemented in a different way, without changing signature, and that, if the changes are going to stay this way, developers that are not you, or officially endorsed by you should be made aware of it explicitly. Your development blog is the perfect place for this, it should not take long to write a post about this topic.

Out of curiosity, I wonder what a developer should reply to a customer who reports warnings after the shipping plugin(s) will have been updated for WC 2.6 and will start throwing warnings in WC 2.5 and earlier (there is a high probability that this will happen, as the WC community has a large percentage of late adopters). I already know that some answers won't cut it:

- **_Update to WC 2.6_**. Not always possible, as some plugins or customisations may not work with WC 2.6. Shifting the blame on their developers for "not being up to speed" is a cop out. Besides, plugins should not require "_always the latest version of WooCommerce_" anyway (it should be obvious why). 
- **_Don't update the shipping plugin_**. Not a good idea, as the updates may be required to fix other issues.
- **_Disable warnings_**. Also not a good idea. Warnings are not supposed to be disabled, they must be addressed. Besides, they are enabled in production quite often, and they can cause all sort of issues (like breaking Ajax calls). That's why warnings and notices are less than ideal for deprecation messages, although I can understand why you used them that way.

After dealing with thousands of customers, I know that most of them would not accept any of those answers as a "solution".
The proper solution would be to ensure that a plugin will work with WC 2.6 as well as 2.5. If the core of those versions are incompatible, in terms of signatures, then plugins will have to handle them differently. As I mentioned, there are ways to do that, but there's a bit of work to be done for that purpose. This is the reason why I consider it important that these changes are clearly announced, and not just left in the documentation (developers may not notice that the documentation has changed).

Bottom line
--
This is a change that could cause issues, and I strongly believe it should be made prominent.

**Comment by daigo75** [comment]
Note for the readers, without opening another issue. WC_Settings_API::process_admin_options()` changes signature as well.

WC 2.5 and earlier
`WC_Settings_API::process_admin_options()`

**Comment by mikejolley** [comment]
Methods cannot remain static forever. The other alternative is to make new methods and use them but a) that's messy b) it introduces more legacy and c) it can conflict with custom methods in the extending class anyway. 

As long as we're not triggering fatal errors, notices are acceptable. As I said it's no different to a deprecated notice which WP uses. No production site should run with WP_DEBUG on to see them.

**Comment by mikejolley** [comment]
And what more, you're basically saying we need to handhold anyone extending a core class. No. That's what beta is for - there is no excuse for not testing against beta/rc if you're extending core. Same as we as WP plugin developers need to test against WP updates.

**Comment by daigo75** [comment]
Sorry, but the assumption of warnings being disabled is optimistic at best, and not a solution. As I wrote, warnings should be addressed, not ignored. This is not just my personal opinion, it's commonly known as a good practice.

As for the fact that methods that cannot remain static, I agree, but core methods are the pillars of a system and they should change as little as possible. The change in `WC_Settings_API` can have widespread repercussions, as it's effectively an "immediate deprecation" of a previous logic, without warning (no pun intended) or fallback. Any plugin using that API will throw warnings in WC 2.6, and it will throw warnings in WC 2.5 as soon as it's updated to work in WC 2.6.
If you have a suggestion to fix that warning (not just hide it), I'm sure that developers will welcome it.

We can "agree to disagree" about this topic, but the issue will remain, and I still think it's important that developers are actively made aware of it. The "_you should test and figure it out_", I'm afraid that it's not a very good answer, especially when the changes affect the core and make it complicated to maintain backward compatibility. 

Writing an announcement is not "holding hands", it's simply informing that the core now has significant changes. If you really don't have time to go into the detail of all changes, writing a simple post to draw the attention to them, and indicating `WC_Settings_API::process_admin_options()` as an example would already help.

I'm already informing several developers with whom I work regularly, but my reach is limited, and that's why I thought of your development blog to reach a wider audience.

Once again, I want to clarify two important points:
- I opened this issue for the benefit of other developers. I'm not affected by these changes and I won't have difficulties dealing with them. 
- I'm not saying that the changes are wrong, I'm just highlighting possible issues and how to prevent them. If anything, I'm trying to help.

Perhaps I gave you the impression that this whole discussion is about the fact that "_I personally don't like the changes_", but that's most definitely not the case. This is simply my feedback, based on my experience and real world scenarios. You are a capable team, I will let you decide what to do with it.

**Note**
I found the discussion where I first read about the reclassification of the signature mismatch in PHP 7: [PHP RFC: Reclassify E_STRICT notices](https://wiki.php.net/rfc/reclassify_e_strict#signature_mismatch_during_inheritance).

From the article:
> If the same signature mismatch occurs when implementing an interface or an abstract function a fatal error is thrown instead of a strict standards notice. **A signature mismatch is a significant issue, which will likely prevent the use of the child object in place of the parent object**.

This may or may not be the case with WC 2.6, but it's still worth keeping in mind that there is a reason if they increased the severity level of a signature mismatch, and not just for it to be ignored by disabling warnings.

**Comment by daigo75** [comment]
All the above is with my most positive intentions, just to be clear. :smile:

**Comment by daigo75** [comment]
@mikejolley as an exercise, I tweaked the changes made to support shipping zones, for the purpose of illustrating how, in my opinion, the same feature could be handled while maintaining backward compatibility. It's quite simple, and the amount of extra code is minimal.

If you would like to have a look, during a :coffee: break, you can find my changes here: https://github.com/aelia-co/woocommerce/tree/POC-Restored_Backward_Compatibility (POC stands for "Proof of Concept").
I haven't had the time to fully test everything, as I wrote the patch in half an hour, but the logic should be clear.

Changes summary
--
- The original methods' signature have been restored, while still maintaining the new functionalities. The new data added in WC 2.6 is set using the parameters that the methods already support.
- There are two new methods to set POST data for the `WC_Settings` class (which is actually only done in **one** place). This is clean and doesn't introduce "legacy", as those are fairly standard getters and setters.

**Comment by mikejolley** [comment]
The settings one may work, however it does open things up to error e.g. if you process two forms and set post data on the first but not the second.

**Comment by daigo75** [comment]
Actually, the logic is the same as before, just replaced by a getter and setter. The getter has a safeguard check too, if no data is explicitly passed it will take the $_POST array.

**Comment by mikejolley** [comment]
Can you throw up a PR for that part only if you want props?

**Comment by daigo75** [comment]
I'm not sure what "props" mean, but I can certainly create a PR if you wish.
Please just keep in mind that mine is still a proof of concept which I threw together quickly, and there may still be room for improvement.

**Comment by mikejolley** [comment]
![2016-04-20 at 09 57](https://cloud.githubusercontent.com/assets/90977/14668902/55498c6c-06de-11e6-94c2-60f3ffe7b08a.png)

B)

**Comment by daigo75** [comment]
I'm still not sure I understand the meaning of the original sentence, but it doesn't matter. :smile: 

I can create a PR, as you suggested, for your review. I don't expect it to be merged as it is, or even to be merged at all, my purpose is just to illustrate a concept.

**Comment by bahiirwa** [comment]
Can I reference this [official-custom-shipping-class-doesnt-work-anymore-shipping-method-api](https://wordpress.org/support/topic/official-custom-shipping-class-doesnt-work-anymore-shipping-method-api/) here? Shipping methods are broken with error message 
`::calculate_shipping($package) should be compatible with WC_Shipping_Method::calculate_shipping($package = Array)`

**Comment by bahiirwa** [comment]
Can I reference this [official-custom-shipping-class-doesnt-work-anymore-shipping-method-api](https://wordpress.org/support/topic/official-custom-shipping-class-doesnt-work-anymore-shipping-method-api/) here? Shipping methods are broken with error message 
`::calculate_shipping($package) should be compatible with WC_Shipping_Method::calculate_shipping($package = Array)`

**Comment by claudiosanches** [comment]
@bahiirwa this does not sounds related to this issue, not to mention that this issue has been patched in 2.5.
